// @backstage/plugin-catalog/dist/components/HasSubdomainsCard/HasSubdomainsCard.esm.js
import { RELATION_HAS_PART } from '@backstage/catalog-model';
import React from 'react';
import { RelatedEntitiesCard } from '../RelatedEntitiesCard/RelatedEntitiesCard.esm.js';
import { componentEntityColumns, asComponentEntities } from '../RelatedEntitiesCard/presets.esm.js';
import { catalogTranslationRef } from '../../alpha/translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

function HasSubdomainsCard(props) {
  const { t } = useTranslationRef(catalogTranslationRef);
  const {
    variant = "gridItem",
    tableOptions = {},
    title = t("hasSubdomainsCard.title")
  } = props;
  return /* @__PURE__ */ React.createElement(
    RelatedEntitiesCard,
    {
      variant,
      title,
      entityKind: "Domain",
      relationType: RELATION_HAS_PART,
      columns: componentEntityColumns,
      asRenderableEntities: asComponentEntities,
      emptyMessage: t("hasSubdomainsCard.emptyMessage"),
      emptyHelpLink: "https://backstage.io/docs/features/software-catalog/descriptor-format/#specsubdomainof-optional",
      tableOptions
    }
  );
}

export { HasSubdomainsCard };
