// @backstage/plugin-scaffolder/dist/alpha/extensions.esm.js
import { convertLegacyRouteRef, compatWrapper } from '@backstage/core-compat-api';
import { PageBlueprint, NavItemBlueprint } from '@backstage/frontend-plugin-api';
import React from 'react';
import { rootRouteRef } from '../routes.esm.js';
import CreateComponentIcon from '@material-ui/icons/AddCircleOutline';
import { FormFieldBlueprint } from '@backstage/plugin-scaffolder-react/alpha';

const scaffolderPage = PageBlueprint.make({
  params: {
    routeRef: convertLegacyRouteRef(rootRouteRef),
    defaultPath: "/create",
    loader: () => import('../components/Router/index.esm.js').then((m) => compatWrapper(/* @__PURE__ */ React.createElement(m.Router, null)))
  }
});
const scaffolderNavItem = NavItemBlueprint.make({
  params: {
    routeRef: convertLegacyRouteRef(rootRouteRef),
    title: "Create...",
    icon: CreateComponentIcon
  }
});
const repoUrlPickerFormField = FormFieldBlueprint.make({
  name: "repo-url-picker",
  params: {
    field: () => import('./fields/RepoUrlPicker.esm.js').then((m) => m.RepoUrlPicker)
  }
});

export { repoUrlPickerFormField, scaffolderNavItem, scaffolderPage };
