// @backstage/plugin-scaffolder/dist/alpha/plugin.esm.js
import { convertLegacyRouteRefs } from '@backstage/core-compat-api';
import { createFrontendPlugin } from '@backstage/frontend-plugin-api';
import { rootRouteRef, selectedTemplateRouteRef, scaffolderTaskRouteRef, actionsRouteRef, scaffolderListTaskRouteRef, editRouteRef, registerComponentRouteRef, viewTechDocRouteRef } from '../routes.esm.js';
import { scaffolderApi } from './api.esm.js';
import { scaffolderPage, scaffolderNavItem, repoUrlPickerFormField } from './extensions.esm.js';
import { formFieldsApi } from './api/FormFieldsApi.esm.js';

var plugin = createFrontendPlugin({
  id: "scaffolder",
  routes: convertLegacyRouteRefs({
    root: rootRouteRef,
    selectedTemplate: selectedTemplateRouteRef,
    ongoingTask: scaffolderTaskRouteRef,
    actions: actionsRouteRef,
    listTasks: scaffolderListTaskRouteRef,
    edit: editRouteRef
  }),
  externalRoutes: convertLegacyRouteRefs({
    registerComponent: registerComponentRouteRef,
    viewTechDoc: viewTechDocRouteRef
  }),
  extensions: [
    scaffolderApi,
    scaffolderPage,
    scaffolderNavItem,
    formFieldsApi,
    repoUrlPickerFormField
  ]
});

export { plugin as default };
